rfcs/2021-07-27-7117-vrl-string-interpolation.md

```diff
@@ -0,0 +1,197 @@
+# RFC 7117 - 2021-07-27 - VRL string interpolation
+
+VRL needs a better way to format strings. Currently the only way to do this is
+to concatenate strings, which can get unwieldly.
+
+## Scope
+
+This RFC discusses creating a new string type within VRL that can use template
+literals, also known as string interpolation, to format strings.
+
+## Pain
+
+Currently the way to create strings is through either string concatenation or
+the `join` function.
+
+Syntactically this is unwieldly. It requires extra key presses and the code
+created doesn't necessarily give an instant idea of what the resulting string
+will look like. Thus the true intent behind the code is obfuscated, which can
+result in bugs.
+
+## User Experience
+
+We will be loosely basing our format strings on Pythons [f-strings](https://datagy.io/python-f-strings/).
+
+To format a string there will be a string type denoted with the prefix `f'`.
+
+Within that string, it is possible to embed VRL expressions by surrounding them
+with `{..}`. VRL will evaluate the expressions and will call `to_string` on
+that expression to return the text representation.
+
+```coffee
+f'The message is { .message } created at { .timestamp }'
+```
+
+If you wish to actually insert a `{` into a string, a double '{{' will be needed.
+
+```coffee
+f'Here is a curly brace -> {{'
+```
+
+Since this is a new string type there are no backward compatibility issues.
+
+### Format strings
+
+The format can be specified by adding format strings after a `:` in the string.
+
+For example to format date the following would be valid:
+
+```coffee
+ts = t'2020-10-21T16:00:00Z'
+f'The time is {ts : %v %R}'
+```
+
+### Errors
+
+The question arises about what we should do if an error occurs.
+
+I would argue that we do not want an f string to be fallible as this would
+cumbersome to the experience of using VRL. If an error occurs the error text is
+output. For example:
+
+```coffee
+f'This is some json { parse_json(.thing) }'
+# This is some json function call error for "parse_json" at (0:18): unable to parse json: expected ident at line 1 column 2
+```
+
+Another source of error would be if the format string is specified for a different
+type - for example using date format strings when the type is an integer.
+
+If this occurs, we just ignore the format string and display the unformatted value.
+
+```coffee
+thing = 2
+f'The date is {thing: %v %R}.'
+# The date is 2.
+```
+
+## Implementation
+
+This new string type can be considered as syntactic sugar for string
+concatenation.
+
+The VRL parser will take a template literal string such as:
+
+```
+i'The message is {{ .message }} created at {{ .timestamp: %v %R }}'
+```
+
+and create an AST identical to the AST for the following expression:
```

Comment:
As this mentions AST specifically... do we need to consider how the VM will handle this or is all good?

Reply:
No. Even in the VM an AST is created. There is just a further step in the VM that compiles that AST to bytecode.

None of the changes here would impact the VM.